File: Src/java/engine-fhir/src/test/java/org/opencds/cqf/cql/engine/fhir/data/Issue1226.java
package org.opencds.cqf.cql.engine.fhir.data;

import static org.junit.Assert.assertEquals;

import java.util.Collections;

import org.hl7.fhir.r4.model.Patient;
import org.hl7.fhir.r4.model.MedicationRequest;
import org.hl7.fhir.r4.model.Reference;
import org.opencds.cqf.cql.engine.data.CompositeDataProvider;
import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider;
import org.opencds.cqf.cql.engine.runtime.Code;
import org.opencds.cqf.cql.engine.runtime.Interval;
import org.testng.annotations.Test;

// https://github.com/cqframework/clinical_quality_language/issues/1226
public class Issue1226 extends FhirExecutionTestBase {

    @Test
    public void medicationReferenceFound() {
        var r = new RetrieveProvider() {
            @Override
            public Iterable<Object> retrieve(String context, String contextPath, Object contextValue, String dataType,
                    String templateId, String codePath, Iterable<Code> codes, String valueSet, String datePath,
                    String dateLowPath, String dateHighPath, Interval dateRange) {

                switch(dataType) {
                    case "Patient" : return Collections.singletonList(new Patient().setId("123"));
                    case "MedicationRequest": return Collections.singletonList(
                        new MedicationRequest()
                            .setMedication(
                                new Reference("Medication/456")));
                }

                return Collections.emptyList();
            }
        };

        var engine = getEngine();

        engine.getState().getEnvironment()
            .registerDataProvider(
                "http://hl7.org/fhir",
                new CompositeDataProvider(r4ModelResolver, r));

        var result = engine.evaluate("Issue1226")
            .forExpression("Most Recent Medication Request reference")
            .value();

        assertEquals("Medication/456", result);
    }

}

File: Src/java/engine-fhir/src/test/kotlin/org/opencds/cqf/cql/engine/fhir/data/Issue1226.kt
package org.opencds.cqf.cql.engine.fhir.data

import org.hl7.fhir.r4.model.MedicationRequest
import org.hl7.fhir.r4.model.Patient
import org.hl7.fhir.r4.model.Reference
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.data.CompositeDataProvider
import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider
import org.opencds.cqf.cql.engine.runtime.Code
import org.opencds.cqf.cql.engine.runtime.Interval

// https://github.com/cqframework/clinical_quality_language/issues/1226
internal class Issue1226 : FhirExecutionTestBase() {
    @Test
    fun medicationReferenceFound() {
        val r: RetrieveProvider =
            object : RetrieveProvider {
                override fun retrieve(
                    context: String?,
                    contextPath: String?,
                    contextValue: Any?,
                    dataType: String,
                    templateId: String?,
                    codePath: String?,
                    codes: Iterable<Code?>?,
                    valueSet: String?,
                    datePath: String?,
                    dateLowPath: String?,
                    dateHighPath: String?,
                    dateRange: Interval?,
                ): Iterable<Any?> {
                    when (dataType) {
                        "Patient" -> return mutableListOf<Any?>(Patient().setId("123"))
                        "MedicationRequest" ->
                            return mutableListOf<Any?>(
                                MedicationRequest().setMedication(Reference("Medication/456"))
                            )
                    }

                    return mutableListOf()
                }
            }

        engine.state.environment.registerDataProvider(
            "http://hl7.org/fhir",
            CompositeDataProvider(r4ModelResolver, r),
        )

        val result =
            engine
                .evaluate("Issue1226")
                .forExpression("Most Recent Medication Request reference")
                .value()

        Assertions.assertEquals("Medication/456", result)
    }
}
